### PR TITLE
Debug option -d demosaic

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
As there are more new things coming for demosaicers it will be good to have a runtime activated debugging option to optionally show detailed info about this stage of the pipeline instead of the compile time option.

Also the color smoothing option has be changed to be introspection ready.

### CODE REVIEW

File: src/iop/demosaic.c [hunk 1]
```diff
@@ -2871,7 +2878,8 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
   dt_iop_roi_t roo = *roi_out;
   roo.x = roo.y = 0;
   // roi_out->scale = global scale: (iscale == 1.0, always when demosaic is on)
-
+  gboolean info = (darktable.unmuted & DT_DEBUG_DEMOSAIC);
+
```
Comment (johnny-bit) [inline src/iop/demosaic.c:2882]: I *think* this should be combined with `PERF` since you use info to produce perf informations for demosaic

Comment (TurboGit) [reply src/iop/demosaic.c:2882]: There is not only perf data. But this one should be DT_DEBUG_PERF.

File: src/iop/demosaic.c [hunk 2]
```diff
@@ -2871,7 +2878,8 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
   dt_iop_roi_t roo = *roi_out;
   roo.x = roo.y = 0;
   // roi_out->scale = global scale: (iscale == 1.0, always when demosaic is on)
-
+  gboolean info = (darktable.unmuted & DT_DEBUG_DEMOSAIC);
```
Comment (TurboGit) [inline src/iop/demosaic.c:2881]: should be const

File: src/iop/demosaic.c [hunk 3]
```diff
@@ -4963,6 +4970,11 @@ void commit_params(struct dt_iop_module_t *self, dt_iop_params_t *params, dt_dev
       dt_control_log(_("`%s' color matrix not found for 4bayer image!"), camera);
     }
   }
+  if(darktable.unmuted & DT_DEBUG_DEMOSAIC)
```
Comment (TurboGit) [inline src/iop/demosaic.c:4973]: this whole if block is equivalent to:

```
dt_print(DT_DEBUG_DEMOSAIC,  "  committed parameters: method: `%s', smooth %i, green %i, CL %i, tiling %i\n",
      method2string(d->demosaicing_method), d->color_smoothing, d->green_eq, piece->process_cl_ready, piece->process_tiling_ready); 
```

**Review by TurboGit — CHANGES_REQUESTED** [review]
some comments

File: src/iop/demosaic.c (hunk 2 above)
Comment (jenshannoschwalm) [reply src/iop/demosaic.c:2881]: good

File: src/iop/demosaic.c (hunk 3 above)
Comment (jenshannoschwalm) [reply src/iop/demosaic.c:4973]: sure

**Comment by jenshannoschwalm** [comment]
BTW, there is some to hint at: in #7756 @kofa73 reports the non-consistant output. It is in fact - as commented there - not an rcd issue but the committed parameters are accepted but no recalculation of the pipeline is triggered.

ATM i don't know how to fix that but using this option can demonstrate the problem ...

**Comment by jenshannoschwalm** [comment]
@TurboGit ok this way?

**Comment by kofa73** [comment]
Is there anything I can do to help test this?

**Review by TurboGit — COMMENTED** [review]
Looks good to me now! Thanks.

**Comment by jenshannoschwalm** [comment]
@kofa73 i think you could open a new issue better describing what the problem is or edit the title of that issue as it not rcd related but just can be found there. try out the new debug option yourself ...

**Comment by jenshannoschwalm** [comment]
@TurboGit , i have never touched/looked at the code managing pipe caching. Can you suggest a starting point to look at, i would like to track this, ...

**Comment by TurboGit** [comment]
@jenshannoschwalm : I have tweaked it a bit but not enough to give you a proper guidance. What I know is that most of the code to look at is in pixelpipe_hb.c, the caching is done via hash values and a mismatch should trigger a refresh. That's basically it :( Each time I looked at this it was for very simple things and I found my way into the code. I think @ralfbrown knows this part far better than me at this stage.

**Comment by jenshannoschwalm** [comment]
Ok and thanks for a starting point; i will do a crawl job and see how far i get. Anyway it's a new section of code i am going to see :-)

**Comment by ralfbrown** [comment]
> What I know is that most of the code to look at is in pixelpipe_hb.c, the caching is done via hash values and a mismatch should trigger a refresh. That's basically it :( 

That's about all I really know about it myself, other than that the code itself is really clunky and  needs an overhaul.

**Pull request merged**